### DISCUSSION TITLE
Stress interpolation

### DISCUSSION THREAD

**Discussion by jessecarterMOOSE** [idaholab/moose#24713] [Q&A Modules: Solid mechanics] [answered]
What methods are available in MOOSE to interpolate stress (or any qp quantity) over the element? Interested in on-the-fly evaluation and saving to the mesh file for postprocessing.

**Comment by ngrilli** [comment]
@jessecarterMOOSE  you could use postprocessors such as the ElementAverageValue, that can average the stress over qp
see example here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/CPDamage/damage_growth.i

postprocessors are evaluated on the fly and can be used to change, for example, the behavior of a material.

**Reply by jessecarterMOOSE** [reply]
Need to keep a little more fidelity than that.

**Reply by GiudGiud** [reply]
by interpret, are you ok with just outputting?

You can output element averages with the https://mooseframework.inl.gov/docs//site/source/auxkernels/MaterialRealAux.html

if you select a different FE family (still elemental) or order, you can get a closer projection

**Comment by jessecarterMOOSE** [comment]
Guess that should have said "interpolate" across the element. Monomials are only approximate, right? Is it least squares fitting? Can the nonlinear variable shape functions be used somehow? 

Then what happens when these discontinuous elemental variables get written to the Exodus file? Is there a different mesh format that will save the raw qp values? 

Just outputting for each element would be useful too for debugging (something like an Abaqus report file)

**Reply by GiudGiud** [reply]
I'm not sure it will let you but if you use the MaterialXXXAux with a different shape function it will do a projection in the backend which amounts to least square fitting. You can look in AuxKernel.C at `AuxKernelTempl<ComputeValueType>::compute()` to see what it does.

It calculates the value of the auxkernel at each qp then tries to fit the auxvariable to those values.

It wont save Qp values.
I dont think we ever save Qp values. We save DOF values.

If you want Qp values your best bet is to modify this VPP to do Qps instead of elements, and for tensor material properties instead of Real ones
https://mooseframework.inl.gov/source/vectorpostprocessors/MaterialVectorPostprocessor.html

**Reply by jessecarterMOOSE** [reply]
How does it reconcile the different nodal values computed by neighboring elements?

**Comment by hugary1995** [comment, marked as answer]
We have this nodal patch recovery system for this exact purpose. It "interpolates" over the element (more precisely over a patch of neighboring elements). I am again responsible for the lack of documentation on this :/

**Reply by jessecarterMOOSE** [reply]
@hugary1995 this? https://mooseframework.inl.gov/help/finite_element_concepts/nodal_patch_recovery.html

**Reply by hugary1995** [reply]
That's the theory, yes. The new system consists of two types of objects:
1. `NodalPatchRecoveryBase` derived userobjects, e.g., `NodalPatchRecoveryMaterialProperty`, which construct the patch for least squares fitting.
2. A "driver" object that actually solves the least squares problem and performs the interpolation. See for example `NodalPatchRecoveryAux`. More specifically, the interpolation happens on this line: https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/auxkernels/NodalPatchRecoveryAux.C#L57, note that `*_current_node` can be replaced with any point inside the patch (the current element and its point neighbors) to interpolate the property.
